File: Backtracking_and_Search/LC_79_Word_Search.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.19

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        m = board.size(), n = board[0].size();
        for (int row = 0; row < m; row++){
            for (int col = 0; col < n; col++){
                if (dfs(board, row, col, 0, word)){
                    return true;
                }
            }
        }
        return false;
    }
private:
    int m, n;
    bool dfs(vector<vector<char>>& board, int row, int col, int i, string word){
        if (i == word.size()){
            return true;
        }
        if (row < 0 || row == m || col < 0 || col == n || board[row][col] != word[i]){
            return false;
        }
        char c = board[row][col];
        board[row][col] = '#';
        bool l = false, r = false, t = false, b = false;
        l = dfs(board, row, col - 1, i+1, word);
        r = dfs(board, row, col + 1, i+1, word);
        t = dfs(board, row - 1, col, i+1, word);
        b = dfs(board, row + 1, col, i+1, word);
        board[row][col] = c;
        return l || r || t || b;
    }
};

File: Backtracking_and_Search/LC_79_Word_Search.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.19

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])

        def dfs(board, row, col, idx):
            if idx == len(word):
                return True
            if row < 0 or row == m or col < 0 or col == n or board[row][col] != word[idx]:
                return False
            temp = board[row][col]
            board[row][col] = "#"
            t = dfs(board, row-1, col, idx+1)
            b = dfs(board, row+1, col, idx+1)
            l = dfs(board, row, col-1, idx+1)
            r = dfs(board, row, col+1, idx+1)
            board[row][col] = temp
            return t or b or l or r


        for row in range(m):
            for col in range(n):
                if dfs(board, row, col, 0):
                    return True
        return False